mathjax3-ts/core/DOMAdaptor.ts

```diff
@@ -0,0 +1,552 @@
+     */
+    getElements(nodes: (string | N | N[])[], document: D): N[];
+
+    /*
+     * @param{N|T} node  The HTML node whose parent is to be obtained
+     * @return{N}        The parent node of the given one
+     */
+    parent(node: N | T): N;
+
+    /*
+     * @param{N} node     The HTML node to be appended to
+     * @param{N|T} child  The node or text to be appended
+     * @return{N|T}       The appended node
+     */
+    append(node: N, child: N | T): N | T;
+
+    /*
+     * @param{N|T} nchild  The node or text to be inserted
+     * @param{N|T} ochild  The node or text where the new child is to be added before it
+     */
+    insert(nchild: N | T, ochild: N | T): void;
+
+    /*
+     * @param{N|T} child  The node or text to be removed from its parent
+     * @return{N|T}       The removed node
+     */
+    remove(child: N | T): N | T;
+
+    /*
+     * @param{N|T} nnode  The node to replace with
+     * @param{N|T} onode  The child to be replaced
+     * @return{N|T}       The removed node
+     */
+    replace(nnode: N | T, onode: N | T): N | T;
+
+    /*
+     * @param{N} node   The HTML node to be cloned
+     * @return{N}       The copied node
+     */
+    clone(node: N): N;
+
+    /*
+     * @param{T} node    The HTML text node to be split
+     * @param{number} n  The index of the character where the split will occur
+     */
+    split(node: T, n: number): T;
+
+    /*
+     * @param{N|T} node   The HTML node whose sibling is to be obtained
+     * @return{N|T}       The node following the given one (or null)
+     */
+    next(node: N | T): N | T;
+
+    /*
+     * @param{N|T} node   The HTML node whose sibling is to be obtained
+     * @return{N|T}       The node preceding the given one (or null)
+     */
+    previous(node: N | T): N | T;
+
+    /*
+     * @param{N} node   The HTML node whose child is to be obtained
+     * @return{N|T}     The first child of the given node (or null)
+     */
+    firstChild(node: N): N | T;
+
+    /*
+     * @param{N} node   The HTML node whose child is to be obtained
+     * @return{N}       The last child of the given node (or null)
+     */
+    lastChild(node: N): N | T;
+
+    /*
+     * @param{N} node    The HTML node whose children are to be obtained
+     * @return{(N|T)[]}  Array of children for the given node (not a live list)
+     */
+    childNodes(node: N): (N | T)[];
+
+    /*
+     * @param{N} node    The HTML node whose child is to be obtained
+     * @param{number} i  The index of the child to return
+     * @return{N|T}      The i-th child node of the given node (or null)
+     */
+    childNode(node: N, i: number): N | T;
+
+    /*
+     * @param{N | T} node   The HTML node whose tag or node name is to be obtained
+     * @return{string}      The tag or node name of the given node
+     */
+    kind(node: N | T): string;
+
+    /*
+     * @param{N|T} node  The HTML node whose value is to be obtained
+     * @return{string}   The value of the given node
```

Comment:
That could also be null. E.g., nodeValue of a non-text node.

Reply:
It looks like `value()` is only being called on text nodes, so there are a couple of possible solutions:

* I could change the parameter to be just `T` rather than `N | T` (though I think that will require casting to T in several places).
* I could change the implementation in the `HTMLAdaptor` to return `node.nodeValue || ''` so that it is never null.
* I could change the return type to `@return{string|null}`.

Do you have a preference?

Reply:
I ended up making it always return a string (no matter whether it was a non-text node).  It turned out this was the most general solution, and the one that required the least fiddly typecasting.

Reply:
Sorry for not replying to the initial remark. Too busy last week. But I like the string solution.